### PR TITLE
Device state retention

### PR DESCRIPTION
**Pull request by benediktibk**
This PR is a follow-up to #71336.

It implements a device state retention module, whose instance is selected via a chosen node. The retention module itself currently stores only the information which device needs to be reinitialized or can be initialized normally. But it should be possible to implement the storage of additional driver specific data as well.

The drivers themselves have to implement the device state retention via a reinit function. The kernel then decides during the initialization if the init or reinit function should be called, dependening on the results from the device state retention module.

This is a deviation from the approach, which was agreed upon in the architecture work group. But I think it will address most of the concerns, as it is a more general approach which can be used for multiple drivers. If it is still undesirable to move forward with this approach another draft is available as #72382 which implements the device state retention only for one specific driver.

### CODE REVIEW

**Review by andyross — COMMENTED** [review]
Some notes.   The device model experts will have more say for sure, but... I guess this seems a little too general?  The problem you're trying to solve is very specific: "don't clobber the bootloader device settings on this GPIO controller", where this is an extremely general framework with a surprising amount (900 lines!) of code added.

Are there any real situations where this is going to help vs. just adding a "use_bootloader_state" flag to the handful of drivers that are sensitive to the transition?

File: include/zephyr/device.h [hunk 1]
```diff
@@ -137,6 +137,47 @@ typedef int16_t device_handle_t;
...
 
+
+/**
+ * @brief Create a device object and set it up for boot time initialization.
+ *
+ * This macro defines a @ref device that is automatically configured by the
+ * kernel during system initialization. This macro should only be used when the
+ * device is not being allocated from a devicetree node. If you are allocating a
+ * device from a devicetree node, use DEVICE_DT_DEFINE() or
+ * DEVICE_DT_INST_DEFINE() instead.
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply include/zephyr/device.h:148]: The documentation definitely has room for improvement, so far I basically only copied the existing stuff.

**Comment by benediktibk** [comment]
> Some notes. The device model experts will have more say for sure, but... I guess this seems a little too general? The problem you're trying to solve is very specific: "don't clobber the bootloader device settings on this GPIO controller", where this is an extremely general framework with a surprising amount (900 lines!) of code added.
> 
> Are there any real situations where this is going to help vs. just adding a "use_bootloader_state" flag to the handful of drivers that are sensitive to the transition?

This approach will solve the transition between bootloader and firmware, as well as the other way round. A "use_bootloader_state" flag won't work in the case of a switch from the firmware back into the bootloader.

The discussion during the architecture work group meeting went a little bit into two directions. Either a more general approach, or a rather specific solution for just one driver. This would be my proposal for the general approach, but I have also prepared the specific one here: #72382

**Comment by bjarki-trackunit** [comment]
We currently have this function in tree
```
bool device_is_ready(const struct device *dev);
```
Which returns true if a device has been initialized successfully. It uses this structure stored in RAM to check:
```
struct device_state {
	/**
	 * Device initialization return code (positive errno value).
	 *
	 * Device initialization functions return a negative errno code if they
	 * fail. In Zephyr, errno values do not exceed 255, so we can store the
	 * positive result value in a uint8_t type.
	 */
	uint8_t init_res;

	/** Indicates the device initialization function has been
	 * invoked.
	 */
	bool initialized : 1;
};
```
Any device can check this at any time, in the existing device init function for example:
```
int pcal64xxa_init(const struct device *dev)
{
        if (device_is_ready(dev)) {
                ...
}
```
It seems way simpler to me use the retention subsystem to preserve the `struct device_state`, instead of adding a new device variation. The retention problem is the same, yet this is compatible with the existing codebase and quite intuitive IMO.

Existing drivers which always reinit don't check if they are already initialized in their init functions, so the addition will not break anything in tree, while providing the ability for new drivers to use it.

Additionally, if retention is not used, and the reinit is not desired, device_is_ready() will always return false in the init function, prompting the init of the driver.

WDYT?

**Comment by benediktibk** [comment]
> We currently have this function in tree
> 
> ```
> bool device_is_ready(const struct device *dev);
> ```
> 
> Which returns true if a device has been initialized successfully. It uses this structure stored in RAM to check:
> 
> ```
> struct device_state {
> 	/**
> 	 * Device initialization return code (positive errno value).
> 	 *
> 	 * Device initialization functions return a negative errno code if they
> 	 * fail. In Zephyr, errno values do not exceed 255, so we can store the
> 	 * positive result value in a uint8_t type.
> 	 */
> 	uint8_t init_res;
> 
> 	/** Indicates the device initialization function has been
> 	 * invoked.
> 	 */
> 	bool initialized : 1;
> };
> ```
> 
> Any device can check this at any time, in the existing device init function for example:
> 
> ```
> int pcal64xxa_init(const struct device *dev)
> {
>         if (device_is_ready(dev)) {
>                 ...
> }
> ```
> 
> It seems way simpler to me to use the retention subsystem to preserve the `struct device_state`, instead of adding a new device variation. The retention problem is the same, yet this is compatible with the existing codebase and quite intuitive IMO.
> 
> Existing drivers which always reinit don't check if they are already initialized in their init functions, so the addition will not break anything in tree, while providing the ability for new drivers to use it.
> 
> Additionally, if retention is not used, and the reinit is not desired, device_is_ready() will always return false in the init function, prompting the init of the driver.
> 
> WDYT?

I like the idea, but I see one problem: There are different types of "initialized". There is the hardware (registers set, ...) and the software (semaphores initialized, ...) point of view. Therefore, reusing the existing flag will create two different meanings for the flag, depending on if the driver for the device calls it or a different part of the code.

But I could store the additional flag (e.g. hardware_initialized) itself in the device_state instead of the separate data structure from the retention module, if this would clear things up a little bit? Of course it then still would be necessary to create another accessor, like device_hardware_is_ready, but this could definitely be a way got get rid of the whole reinit-shenanigans.

**Comment by bjarki-trackunit** [comment]
> I like the idea, but I see one problem: There are different types of "initialized". There is the hardware (registers set, ...) and the software (semaphores initialized, ...) point of view. Therefore, reusing the existing flag will create two different meanings for the flag, depending on if the driver for the device calls it or a different part of the code.

>But I could store the additional flag (e.g. hardware_initialized) itself in the device_state instead of the separate data structure from the retention module, if this would clear things up a little bit? Of course it then still would be necessary to create another accessor, like device_hardware_is_ready, but this could definitely be a way got get rid of the whole reinit-shenanigans.

I see, and agreed, adding a bit for "hardware initialized" to the existing struct is quite intuitive. device_is_ready() must retain its meaning -> both "software" and "hardware" initialized, so adding a new function to check the flag is also required.

maybe `device_is_post_init()`, `device_is_initialized()`?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
I still believe this to be a too generic solution to a very isolated, specific downstream problem.

I am sorry, but I do not see a general need for a device state retention subsystem.

**Pull request closed without merging**